### PR TITLE
Fuzzy Search

### PR DESCRIPTION
**Pull request by kat-hall123**
## Description
- This PR implements the fuzzy search feature. When a user searches for a provider, the results are displayed in an order that best matches the query. Typos like substitutions, deletions, insertions, transpostions are handled in the process. 
- What's Next: Potential Stretch Goals

## Milestones
- TC 2 (Fuzzy Search): Complete

## Test Plan
https://github.com/user-attachments/assets/b768fcf0-5595-4b64-9064-3f8675abe1c8

### CODE REVIEW

File: backend/routes/providers.js [hunk 1]
```diff
@@ -6,25 +6,86 @@ const router = require('express').Router()
...
+    let i = 0
+    let j = 0
+    let prefixLength = 0
+
+    while(i < input.length && i < target.length && input[i] === target[i] && prefixLength < 4) {
+        prefixLength++
+        i++
+    }
+
+    let bonus = prefixLength * 0.05
```
Comment (maurenej) [inline backend/routes/providers.js:50]: Nit: Put 0.05 into a global constant eg. (`BONUS_MULTIPLIER`)

This makes it so it can be reused later (though I don't expect that to be needed in your case) and only needs to be modified in one place
```suggestion
    let bonus = prefixLength * BONUS_MULTIPLIER
```

File: backend/routes/providers.js [hunk 2]
```diff
@@ -6,25 +6,86 @@ const router = require('express').Router()
...
+
+    let cost = 0
+    i = j = prefixLength
```
Comment (maurenej) [inline backend/routes/providers.js:53]: Actually, now thinking about this, could we be missing out on more transpositions/insertions/deletions by in the prefix since we only checked the prefix for equality?

```
// Target
K a t h e r i n e

A k h t e r i n e 
P p p p e r i n e
```
Should these two inputs be scored differently?

What about
```
// Target
K a t h e r i n e

A t h e r i n e
P p p p r i n e
```

File: backend/routes/providers.js [hunk 3]
```diff
@@ -6,25 +6,86 @@ const router = require('express').Router()
...
+
+    while(i < input.length && j < target.length) {
```
Comment (maurenej) [inline backend/routes/providers.js:55]: I think you can check `i + 1 < input.length` and `j + 1 < target.length` here (instead of in the `if` statements) and any additional characters would be caught by line 81

File: backend/routes/providers.js [hunk 4]
```diff
@@ -6,25 +6,86 @@ const router = require('express').Router()
...
+        } else {
+            // Substitution
+            cost += 1
+            i++
+            j++
+        }
+    }
+
+    // If there were still remaining characters not compared
+    cost += (input.length - i) + (target.length - j)
```
Comment (maurenej) [inline backend/routes/providers.js:81]: Thinking a bit about the costs--would these give the same costs? Is that ok?
```
// Target
K a t h e r i n e

C k k k t t h h e e r r i i n n e e
Y y y y y y y y y y y y y y y y y y
```

Basically, with insertions and deletions, you are able to match a letter. Is that more similar than not matching anything in a substitution? (Also no right or wrong--just wondering if there should be priority to having the same letters)

**Review by maurenej — CHANGES_REQUESTED** [review]
Very cool test plan video! Added a few edge cases that I thought of when looking at the algorithm. Take a look and decide which inputs in these cases should be ranked higher in your opinion!

File: backend/routes/providers.js (hunk 2 above)
Comment (kat-hall123) [reply backend/routes/providers.js:53]: I actually think the prefixes are also checked for for transpositions/insertions/deletions. For A k h t e r i n e, because the length of prefixes are 0, the comparison loop is able to start from index 0 as well, so A k h t e r i n e goes through the transposition check. These are what I get as results:
Screenshot 2025-07-18 at 10 07 56 AM
Screenshot 2025-07-18 at 10 08 10 AM
Screenshot 2025-07-18 at 10 08 25 AM
Screenshot 2025-07-18 at 10 08 38 AM
I feel like these make sense as the results?

File: backend/routes/providers.js (hunk 4 above)
Comment (kat-hall123) [reply backend/routes/providers.js:81]: Yeah, this is a good point. I think I'll change the insertion/deletion costs to only 0.8 while keeping substitution at 1. In this way, a word that is misaligned is "more similar" than a word that fully did not match a character. I feel like this is probably more user friendly. Essentially, "Kjatherine" would be better than "Kjtherine" right?
